### DISCUSSION TITLE
DDP and InputNormalization

### DISCUSSION THREAD

**Discussion by lucadellalib** [speechbrain/speechbrain#2381] [Feature Request]
### 🚀 The feature

`speechbrain.processing.features.InputNormalization` implementation is unusual compared to PyTorch built-in normalizations as it does not register any buffer (see for example https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/batchnorm.py#L54). Besides requiring custom `to`  and save/load hooks implementation, it might lead to inconsistencies when training with DDP. By default DDP broadcasts the value of each buffer from the rank 0 process to the others (see argument `broadcast_buffers` in https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html) , however `InputNormalization` does not register any buffer. Therefore, running statistics are local to each process independently of the value of `broadcast_buffers`.

### Solution outline

Consider refactoring the implementation of `InputNormalization` using regular PyTorch buffers (however this might lead to incompatibilities with previously trained models).

**Comment by mravanelli** [comment]
Hi,
interesting observation. Could you please come out with a PR with a fix?

**Comment by gupnaval** [comment]
I got a similar issue when I trained a model on two different versions namely 0.5.11 and 0.5.14. The output models of version 0.5.14 doesn't contained statistics like **norm.running_mean** and **norm.running_var**. 

I have posted this issue in the bug section also.

**Comment by TParcollet** [comment]
Hi, this is most likely due to the fact that the normalisation technique has changed from one that has running parameters (like global normalisation) against sentence normalisation. This change should be visible in the yaml.

**Comment by gupnaval** [comment]
I completely got you TParcollet, but here I have used the same **train.py** and **train.yaml** files for training in both versions of speechbrain namely 0.5.11 and 0.5.14. 
I am a little new to this area, so don't know the complete basics. Kindly bear with me and if you can some insightful knowledge or links where I can understand these things better, I will be grateful
